File: frontend/apps/ios/Pantopus/Core/Networking/Models/AudienceProfile/AudienceProfileDTOs.swift
//
//  AudienceProfileDTOs.swift
//  Pantopus
//
//  Decoder shapes for /api/personas/* — the owner-facing audience
//  dashboard (T3.3). Backend column names stay (persona / fan /
//  broadcast / tier); the UI renames at the VM boundary.
//

import Foundation

// MARK: - GET /api/personas/me

/// Envelope from `GET /api/personas/me` — the persona summary + its
/// primary broadcast channel ID needed by the composer.
public struct PersonaMeResponse: Decodable, Sendable {
    public let persona: PersonaSummaryDTO?
    public let channel: BroadcastChannelDTO?
}

public struct PersonaSummaryDTO: Decodable, Sendable, Hashable {
    public let id: String
    public let handle: String?
    public let displayName: String?
    public let avatarUrl: String?
    public let bannerUrl: String?
    public let bio: String?
    public let category: String?
    public let audienceLabel: String?
    public let followerCount: Int?
    public let postCount: Int?

    enum CodingKeys: String, CodingKey {
        case id, handle, bio, category
        case displayName, avatarUrl, bannerUrl
        case audienceLabel
        case followerCount, postCount
    }
}

public struct BroadcastChannelDTO: Decodable, Sendable, Hashable {
    public let id: String
    public let title: String?
    public let description: String?
    public let status: String?

    enum CodingKeys: String, CodingKey {
        case id, title, description, status
    }
}

// MARK: - GET /api/personas/me/audience

public struct AudienceListResponse: Decodable, Sendable {
    public let persona: PersonaSummaryDTO?
    public let items: [FanDTO]
    public let counts: AudienceCountsDTO
}

public struct AudienceCountsDTO: Decodable, Sendable, Hashable {
    public let totalActive: Int?
    public let pending: Int?
    public let byTier: [String: Int]?

    enum CodingKeys: String, CodingKey {
        case totalActive, pending, byTier
    }
}

/// One follower row in the creator's audience list — strictly the
/// creator-side serializer (no PII per the privacy gate). Backend
/// returns camelCase keys here (`fanHandle` not `fan_handle`), so no
/// per-field CodingKeys are needed.
public struct FanDTO: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let membershipId: String?
    public let fanHandle: String?
    public let fanDisplayName: String?
    public let fanAvatarUrl: String?
    public let status: String?
    public let tier: FanTierBadgeDTO?
    public let verifiedLocal: Bool?
    public let joinedMonth: String?
    public let tenureMonths: Int?
    public let cancelAtPeriodEnd: Bool?

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        // Server emits `membershipId` (no separate `id`); map either
        // key to our `id` so the serializer evolution stays compatible.
        let mid = try container.decodeIfPresent(String.self, forKey: .membershipId)
        let raw = try container.decodeIfPresent(String.self, forKey: .id)
        id = mid ?? raw ?? ""
        membershipId = mid
        fanHandle = try container.decodeIfPresent(String.self, forKey: .fanHandle)
        fanDisplayName = try container.decodeIfPresent(String.self, forKey: .fanDisplayName)
        fanAvatarUrl = try container.decodeIfPresent(String.self, forKey: .fanAvatarUrl)
        status = try container.decodeIfPresent(String.self, forKey: .status)
        tier = try container.decodeIfPresent(FanTierBadgeDTO.self, forKey: .tier)
        verifiedLocal = try container.decodeIfPresent(Bool.self, forKey: .verifiedLocal)
        joinedMonth = try container.decodeIfPresent(String.self, forKey: .joinedMonth)
        tenureMonths = try container.decodeIfPresent(Int.self, forKey: .tenureMonths)
        cancelAtPeriodEnd = try container.decodeIfPresent(Bool.self, forKey: .cancelAtPeriodEnd)
    }

    enum CodingKeys: String, CodingKey {
        case id, membershipId, fanHandle, fanDisplayName, fanAvatarUrl
        case status, tier, verifiedLocal, joinedMonth, tenureMonths, cancelAtPeriodEnd
    }
}

public struct FanTierBadgeDTO: Decodable, Sendable, Hashable {
    public let rank: Int?
    public let name: String?
}

// MARK: - GET /api/personas/:handle/posts

public struct PersonaPostsResponse: Decodable, Sendable {
    public let posts: [PersonaPostDTO]
}

public struct PersonaPostDTO: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let body: String?
    public let createdAt: String?
    public let visibility: String?
    public let targetTierRank: Int?
    public let deliveredCount: Int?
    public let readCount: Int?
    public let mediaUrls: [String]?

    enum CodingKeys: String, CodingKey {
        case id, body, visibility
        case createdAt = "created_at"
        case targetTierRank = "target_tier_rank"
        case deliveredCount = "delivered_count"
        case readCount = "read_count"
        case mediaUrls = "media_urls"
    }
}

// MARK: - GET /api/personas/:handle/tiers

public struct PersonaTiersResponse: Decodable, Sendable {
    public let tiers: [PersonaTierDTO]
}

public struct PersonaTierDTO: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let rank: Int
    public let name: String
    public let description: String?
    public let priceCents: Int?
    public let currency: String?
}

// MARK: - GET /api/personas/:id/membership-stats

public struct MembershipStatsResponse: Decodable, Sendable {
    public let counts: MembershipStatsCounts
}

public struct MembershipStatsCounts: Decodable, Sendable, Hashable {
    public let followers: Int?
    public let members: Int?
    public let insiders: Int?
    public let direct: Int?
}

// MARK: - GET /api/personas/:id/dms/threads

public struct PersonaThreadsResponse: Decodable, Sendable {
    public let threads: [PersonaThreadDTO]
}

public struct PersonaThreadDTO: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let membershipId: String?
    public let fanHandle: String?
    public let fanDisplayName: String?
    public let fanAvatarUrl: String?
    public let tier: FanTierBadgeDTO?
    public let lastMessagePreview: String?
    public let lastMessageAt: String?
    public let unreadCount: Int?
    public let status: String?
}

// MARK: - POST /api/broadcast/channels/:id/messages

public struct PublishUpdateResponse: Decodable, Sendable {
    public let message: BroadcastMessageDTO?
}

public struct BroadcastMessageDTO: Decodable, Sendable, Hashable {
    public let id: String?
    public let body: String?
    public let visibility: String?
    public let targetTierRank: Int?
    public let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id, body, visibility
        case targetTierRank = "target_tier_rank"
        case createdAt = "created_at"
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/audience/AudienceProfileDtos.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.data.api.models.audience

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

// GET /api/personas/me

@JsonClass(generateAdapter = true)
data class PersonaMeResponse(
    val persona: PersonaSummaryDto? = null,
    val channel: BroadcastChannelDto? = null,
)

@JsonClass(generateAdapter = true)
data class PersonaSummaryDto(
    val id: String,
    val handle: String? = null,
    val displayName: String? = null,
    val avatarUrl: String? = null,
    val bannerUrl: String? = null,
    val bio: String? = null,
    val category: String? = null,
    val audienceLabel: String? = null,
    val followerCount: Int? = null,
    val postCount: Int? = null,
)

@JsonClass(generateAdapter = true)
data class BroadcastChannelDto(
    val id: String,
    val title: String? = null,
    val description: String? = null,
    val status: String? = null,
)

// GET /api/personas/me/audience

@JsonClass(generateAdapter = true)
data class AudienceListResponse(
    val persona: PersonaSummaryDto? = null,
    val items: List<FanDto> = emptyList(),
    val counts: AudienceCountsDto = AudienceCountsDto(),
)

@JsonClass(generateAdapter = true)
data class AudienceCountsDto(
    val totalActive: Int? = null,
    val pending: Int? = null,
    val byTier: Map<String, Int>? = null,
)

/**
 * One follower row. The backend's creator-side serializer emits
 * camelCase keys (`fanHandle`, not `fan_handle`) so no `@Json` mapping
 * is needed. `id` is sourced from `membershipId` when present.
 */
@JsonClass(generateAdapter = true)
data class FanDto(
    val membershipId: String? = null,
    val fanHandle: String? = null,
    val fanDisplayName: String? = null,
    val fanAvatarUrl: String? = null,
    val status: String? = null,
    val tier: FanTierBadgeDto? = null,
    val verifiedLocal: Boolean? = null,
    val tenureMonths: Int? = null,
    val joinedMonth: String? = null,
    val cancelAtPeriodEnd: Boolean? = null,
) {
    val id: String get() = membershipId ?: fanHandle ?: ""
}

@JsonClass(generateAdapter = true)
data class FanTierBadgeDto(
    val rank: Int? = null,
    val name: String? = null,
)

// GET /api/personas/:handle/posts

@JsonClass(generateAdapter = true)
data class PersonaPostsResponse(
    val posts: List<PersonaPostDto> = emptyList(),
)

@JsonClass(generateAdapter = true)
data class PersonaPostDto(
    val id: String,
    val body: String? = null,
    @Json(name = "created_at") val createdAt: String? = null,
    val visibility: String? = null,
    @Json(name = "target_tier_rank") val targetTierRank: Int? = null,
    @Json(name = "delivered_count") val deliveredCount: Int? = null,
    @Json(name = "read_count") val readCount: Int? = null,
    @Json(name = "media_urls") val mediaUrls: List<String>? = null,
)

// GET /api/personas/:handle/tiers

@JsonClass(generateAdapter = true)
data class PersonaTiersResponse(
    val tiers: List<PersonaTierDto> = emptyList(),
)

@JsonClass(generateAdapter = true)
data class PersonaTierDto(
    val id: String,
    val rank: Int,
    val name: String,
    val description: String? = null,
    val priceCents: Int? = null,
    val currency: String? = null,
)

// GET /api/personas/:id/membership-stats

@JsonClass(generateAdapter = true)
data class MembershipStatsResponse(
    val counts: MembershipStatsCountsDto = MembershipStatsCountsDto(),
)

@JsonClass(generateAdapter = true)
data class MembershipStatsCountsDto(
    val followers: Int? = null,
    val members: Int? = null,
    val insiders: Int? = null,
    val direct: Int? = null,
)

// GET /api/personas/:id/dms/threads

@JsonClass(generateAdapter = true)
data class PersonaThreadsResponse(
    val threads: List<PersonaThreadDto> = emptyList(),
)

@JsonClass(generateAdapter = true)
data class PersonaThreadDto(
    val id: String,
    val membershipId: String? = null,
    val fanHandle: String? = null,
    val fanDisplayName: String? = null,
    val fanAvatarUrl: String? = null,
    val tier: FanTierBadgeDto? = null,
    val lastMessagePreview: String? = null,
    val lastMessageAt: String? = null,
    val unreadCount: Int? = null,
    val status: String? = null,
)

// POST /api/broadcast/channels/:id/messages

@JsonClass(generateAdapter = true)
data class PublishUpdateBody(
    val body: String,
    val visibility: String,
    @Json(name = "target_tier_rank") val targetTierRank: Int? = null,
)

@JsonClass(generateAdapter = true)
data class PublishUpdateResponse(
    val message: BroadcastMessageDto? = null,
)

@JsonClass(generateAdapter = true)
data class BroadcastMessageDto(
    val id: String? = null,
    val body: String? = null,
    val visibility: String? = null,
    @Json(name = "target_tier_rank") val targetTierRank: Int? = null,
    @Json(name = "created_at") val createdAt: String? = null,
)
